src/riak_kv_wm_timeseries.erl

```diff
@@ -29,572 +34,382 @@
-              %% body
-              key     :: undefined |  ts_rec(),  %% parsed out of JSON that came in the body
-              data    :: undefined | [ts_rec()], %% ditto
-              result  :: undefined | ok | {Headers::[binary()], Rows::[ts_rec()]}
-             }).
+-record(ctx,
+        {api_call    :: 'undefined' | 'get' | 'put' | 'delete',
+         table       :: 'undefined' | binary(),
+         mod         :: 'undefined' | module(),
+         key         :: 'undefined' | ts_rec(),
+         object,
+         timeout :: 'undefined' | integer(),
+         options,  %% for the call towards riak.
+         prefix,
+         riak}).
 
 -define(DEFAULT_TIMEOUT, 60000).
--define(TABLE_ACTIVATE_WAIT, 30).   %% wait until table's bucket type is activated
 
--define(CB_RV_SPEC, {boolean()|atom()|tuple(), #wm_reqdata{}, #ctx{}}).
+-type cb_rv_spec(T) :: {T, #wm_reqdata{}, #ctx{}}.
+-type halt() :: {'halt', 200..599} | {'error' , term()}.
 -type ts_rec() :: [riak_pb_ts_codec:ldbvalue()].
 
-
 -spec init(proplists:proplist()) -> {ok, #ctx{}}.
 %% @doc Initialize this resource.  This function extracts the
 %%      'prefix' and 'riak' properties from the dispatch args.
 init(Props) ->
     {ok, #ctx{prefix = proplists:get_value(prefix, Props),
               riak = proplists:get_value(riak, Props)}}.
+    %% {{trace, "/tmp"}, #ctx{prefix = proplists:get_value(prefix, Props),
+    %%                        riak = proplists:get_value(riak, Props)}}.
+%% wmtrace_resource:add_dispatch_rule("wmtrace", "/tmp").
 
--spec service_available(#wm_reqdata{}, #ctx{}) ->
-    {boolean(), #wm_reqdata{}, #ctx{}}.
+-spec service_available(#wm_reqdata{}, #ctx{}) -> cb_rv_spec(boolean() | halt()).
 %% @doc Determine whether or not a connection to Riak
-%%      can be established.  This function also takes this
-%%      opportunity to extract the 'table' and 'key' path
-%%      bindings from the dispatch.
-service_available(RD, Ctx = #ctx{riak = RiakProps}) ->
+%%      can be established.
+%%      Convert the table name from the part of the URL.
+service_available(RD, #ctx{riak = RiakProps}=Ctx) ->
     case riak_kv_wm_utils:get_riak_client(
            RiakProps, riak_kv_wm_utils:get_client_id(RD)) of
-        {ok, C} ->
-            {true, RD,
-             Ctx#ctx{api_version = wrq:path_info(api_version, RD),
-                     method = wrq:method(RD),
-                     client = C,
-                     table = utf8_to_binary(
-                               mochiweb_util:unquote(
-                                 wrq:path_info(table, RD)))
-                    }};
+        {ok, _C} ->
+            Table = riak_kv_wm_ts_util:table_from_request(RD),
+            Mod = riak_ql_ddl:make_module_name(Table),
+            {true, RD, Ctx#ctx{table=Table, mod=Mod}};
         {error, Reason} ->
-            handle_error({riak_client_error, Reason}, RD, Ctx)
+            Resp = riak_kv_wm_ts_util:set_error_message("Unable to connect to Riak: ~p", [Reason], RD),
```

Comment:
I don't think the error message reflects the problem - it's that the node is starting or stopping, not that you are unable to connect.